exp/mutators/mutator.go

```diff
@@ -0,0 +1,76 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mutators
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// ResourcesMutator mutates in-place a slice of ASO resources to be reconciled. These mutations make only the
+// changes strictly necessary for CAPZ resources to play nice with Cluster API. Any mutations should be logged
+// and mutations that conflict with user-defined values should be rejected by returning Incompatible.
+type ResourcesMutator func(context.Context, []*unstructured.Unstructured) error
+
+type mutation struct {
+	location string
+	val      any
+	reason   string
+}
+
+func logMutation(log logr.Logger, mutation mutation) {
+	log.V(4).Info(fmt.Sprintf("setting %s to %v %s", mutation.location, mutation.val, mutation.reason))
+}
+
+// Incompatible describes an error where a piece of user-defined configuration does not match what CAPZ
+// requires.
+type Incompatible struct {
+	mutation
+	userVal any
+}
+
+func (e Incompatible) Error() string {
+	return fmt.Sprintf("incompatible value: value at %s set by user to %v but CAPZ must set it to %v %s. The user-defined value must not be defined, or must match CAPZ's desired value.", e.location, e.userVal, e.val, e.reason)
+}
+
+// ApplyMutators applies the given mutators to the given resources.
```

Comment:
From what I see, `ApplyMutators` also converts resources to the unstructured type. I got a bit confused on when this function was called without any mutators such as right here: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/4793/files#diff-422c3ed0c1ff47314c4f39e56d23c8a5f7f5c8ac24dcbb1c5b5da60834f23243R230

Maybe there is a better name for it since there are a few cases where mutators are not being applied.

Reply:
Would it help if there was also a function like `ToUnstructured` that was basically just an alias for this that doesn't take any function parameters?

Reply:
Yeah I think that would make it more clear, but it's just a design choice and I think it's also fine the way it is.

Reply:
Just pushed this.